hnix-store-core/src/System/Nix/Internal/StorePath.hs

```diff
@@ -0,0 +1,109 @@
+{-|
+Description : Representation of Nix store paths.
+-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module System.Nix.Internal.StorePath where
+import System.Nix.Hash (HashAlgorithm(Truncated, SHA256), Digest, encodeBase32)
+import Text.Regex.Base.RegexLike (makeRegex, matchTest)
+import Text.Regex.TDFA.Text (Regex)
+import Data.Text (Text)
+import qualified Data.Text as T
+import Data.Text.Encoding (encodeUtf8)
+import GHC.TypeLits (Symbol, KnownSymbol, symbolVal)
+import System.FilePath ((</>))
+import Data.Proxy (Proxy(Proxy))
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BC
+
+-- | The hash algorithm used for store path hashes.
+type StorePathHashAlgo = 'Truncated 20 'SHA256
+
+-- | Regular expression to match valid store path names.
+storePathNameRegex :: Regex
+storePathNameRegex = makeRegex r
+  where
+    r :: String
+    r = "[a-zA-Z0-9\\+\\-\\_\\?\\=][a-zA-Z0-9\\+\\-\\.\\_\\?\\=]*"
+
+-- | The name portion of a Nix path.
+--
+-- 'unStorePathName' must only contain a-zA-Z0-9+._?=-, can't start
+-- with a -, and must have at least one character (i.e. it must match
+-- 'storePathNameRegex').
+newtype StorePathName = StorePathName
+  { -- | Extract the contents of the name.
+    unStorePathName :: Text
+  }
+
+-- | Smart constructor for 'StorePathName' that ensures the underlying
+-- content invariant is met.
+makeStorePathName :: Text -> Maybe StorePathName
+makeStorePathName n = case matchTest storePathNameRegex n of
+  True  -> Just $ StorePathName n
+  False -> Nothing
+
+-- | A type-level representation of the root directory of a Nix store.
+type StoreDir = Symbol
+
+-- | A path in a Nix store.
+--
+-- @storeDir@: The root of the Nix store (e.g. \/nix\/store).
+data StorePath (storeDir :: StoreDir) = StorePath
+  { storePathHash :: !(Digest StorePathHashAlgo)
+  , storePathName :: !StorePathName
+  }
+
+-- | A 'StoreDir' whose value is known at compile time.
+type KnownStoreDir = KnownSymbol
+
+-- | Render a 'StorePath' as a 'ByteString'.
+--
+-- Would prefer to return a @RawFilePath@ here, but we want to avoid
+-- the dependency on the @unix@ package.
+storePathToByteString
+  :: forall storeDir . (KnownStoreDir storeDir)
+  => StorePath storeDir
+  -> ByteString
+storePathToByteString (StorePath {..}) = root <//> hashPart <//> name
```

Comment:
Should it be `root <//> hashPart <--> name`?

Reply:
Fixed, :man_facepalming: